backend/common/utils/rollup.py

```diff
@@ -32,6 +32,56 @@ def ancestors(cell_type):
     return ancestors
 
 
+def get_valid_descendants(
+    cell_type: str, valid_cell_types: frozenset[str], cell_counts: Optional[dict[str, int]] = None
+):
+    """
+    Get valid descendants for a cell type.
+
+    Terms can be suffixed with ";;{A_0}--{B_0}--..." to indicate that only descendants within the same
+    group specified by the suffix should be considered. For example, if the term is
+    "CL:0000540;;{tissue_0}--{disease_0}" then only descendants with the same tissue_0 and disease_0
+    will be considered.
+
+    Arguments
+    ---------
+    cell_type : str
+        Cell type (cell type ontology term ID, potentially suffixed)
+    valid_cell_types : frozenset[str]
+        FrozenSet of valid cell types
+    cell_counts : dict[str, int], optional, default=None
+        Dictionary mapping cell type ontology term IDs (potentially suffixed) to the number of cells of that type.
+        This is used to exclude all descendants of a cell type if the cell type has the same number of cells as one
+        of its descendants. In these cases, the cell type is a redundant node and should be excluded.
+
+
+    Returns
+    -------
+    list[str] - An empty list indicates that the cell type is a redundant node and should be excluded.
+    """
+    # if the input cell types are suffixed, only consider descendants within the same group
+    if ";;" in cell_type:
+        prefix, suffix = cell_type.split(";;")
+        suffix = ";;" + suffix
+    else:
+        prefix = cell_type
+        suffix = ""
+
+    # find descendants of cell type and re-suffix them
+    relatives = [f"{i}{suffix}" for i in descendants(prefix)]
+    valid_relatives = list(valid_cell_types.intersection(relatives))
+    if cell_counts is not None:
+        relative_cell_counts = [
+            cell_counts.get(relative)
+            for relative in valid_relatives
+            if relative != cell_type and cell_counts.get(relative) is not None
+        ]
+        if cell_counts.get(cell_type) in relative_cell_counts:
```

Comment:
Should we check that `len(relative_cell_counts) == 1` and raise an exception otherwise? Because are now looking for a linear chain here right?

Also a comment here to better explain why we are doing this will be really useful for someon reading this later

Reply:
Why does the length of `relative_cell_counts` have to be 1?

Reply:
I can't think of any trivial assertions we can make that would ensure the sums are calculated properly. That's why we have separate unit tests for the rollup operation.

Reply:
Yea sorry I was mistaken about the assertion that `len(relative_cell_counts) == 1`. 

You could have the following graph (expressed as an adjacency list here):

```
A: B
B: C
C: D
D: [E, F]
```

Here `[A, B, C, D]` form a linear chain. 

Assume _pre-rollup_ numbers of:

```
A: 0
B: 0
C: 0
D: 1
E: 10
F: 2
```

_Unpruned post-rollup_ numbers:

```
A: 13
B: 13
C: 13
D: 13
E: 10
F: 2
```

The descendants of `A` would be `[B, C, D, E, F]`. Then _post-rollup_ the descendant cell-counts list for `A` (ie `relative_cell_counts`) would have entries `[B: 13, C: 13, D: 13, E: 10, F: 2]`, which has length greater than 1. In this case, you would exclude `A, B, C` when processing each in turn